compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -921,13 +917,74 @@ private boolean isXMLTypeAssignable(BType sourceType, BType targetType, Set<Type
                     return isAssignable(sourceType, target.constraint, unresolvedTypes);
                 }
                 BXMLType source = (BXMLType) sourceType;
+                if (source.constraint.tag == TypeTags.NEVER) {
+                    if (sourceTag == targetTag) {
+                        return true;
+                    }
+                    return isAssignable(source, target.constraint, unresolvedTypes);
+                }
                 return isAssignable(source.constraint, target.constraint, unresolvedTypes);
             }
             return true;
         }
+        if (sourceTag == TypeTags.XML) {
+            BXMLType source = (BXMLType) sourceType;
+            if (targetTag == TypeTags.XML_TEXT) {
+                if (source.constraint != null) {
+                    return source.constraint.tag == TypeTags.NEVER;
+                }
+            }
+            if (targetTag == TypeTags.STRING) {
+                if (source.constraint.tag == TypeTags.NEVER) {
+                    return true;
+                }
+                return isAssignable(source.constraint, targetType, unresolvedTypes);
+            }
+        }
+        if (sourceTag == TypeTags.XML_TEXT && targetTag == TypeTags.STRING) {
+            return true;
+        }
         return sourceTag == targetTag;
     }
 
+    public boolean isXMLExprCastableToString(BType target, BType source) {
```

Comment:
Just for consistency with other methods shall we switch the parameters here?
```suggestion
    public boolean isXMLExprCastableToString(BType source, BType target) {
```